HELIX failures investigations

DamianEdwards opened the pull request:
Investigating helix failures.

DamianEdwards committed:
Update template baseline test to use original file name

DamianEdwards committed:
Comment out tenmplate baseline test namespace declaration

DamianEdwards committed:
Comment out template warning checks

DamianEdwards committed:
Trigger matrix helix builds for this branch

dougbu commented:
Suggest adding `CollectDump` to the `--blame` option at https://github.com/dotnet/aspnetcore/blob/main/eng/tools/HelixTestRunner/TestRunner.cs#L224

May also be worthwhile adding `VSTEST_ DISABLE_FASTER_JSON_SERIALIZATION=1` and `DOTNET_CLI_VSTEST_TRACE=1` after `https://github.com/dotnet/aspnetcore/blob/main/eng/tools/HelixTestRunner/TestRunner.cs#L37

DamianEdwards committed:
Enable more tracing & dump collect in helix test runner

DamianEdwards committed:
Disable test parallelization for project template tests

dougbu commented on src/ProjectTemplates/Shared/Project.cs:
```diff
@@ -156,11 +156,12 @@ internal async Task<ProcessResult> RunDotNetPublishAsync(IDictionary<string, str
 
         var result = new ProcessResult(execution);
 
+        // Comment out to see if it impacts test failures
```
nit: Would be really surprising if this tiny bit of code caused the problems we're seeing😉

DamianEdwards replied:
Yeah I know. Just trying to see if any of the test additions are involved in the failures.

dougbu commented on src/ProjectTemplates/test/BaselineTest.cs:
```diff
@@ -102,19 +102,20 @@ public async Task Template_Produces_The_Right_Set_Of_FilesAsync(string arguments
-                    Assert.StartsWith(Project.ProjectName, namespaceDeclaration, StringComparison.Ordinal);
-                }
-            }
+            // Commented out to see if it impacts the Helix test failures
```
Is the plan to restore (uncomment) this block after you get to a successful run❔

DamianEdwards replied:
Yes. I'm just commenting out things that were added just before the failures started happening to see if they're at all involved.

dougbu commented:
/fyi @MarcoRossignoli this PR incorporates your suggestions

DamianEdwards committed:
Update Web API template test to ensure unique project key

DamianEdwards committed:
Remove parallelism from template tests

- Ensure project directory name uniqueness
- Comment out DotNetNewLock for now

DamianEdwards committed:
Fix Blazor WASM template tests

HaoK commented on src/ProjectTemplates/Shared/ProjectFactoryFixture.cs:
```diff
@@ -35,33 +51,36 @@ public async Task<Project> GetOrCreateProject(string projectKey, ITestOutputHelp
+            DiagnosticsMessageSink = DiagnosticsMessageSink,
+            ProjectGuid = Path.GetRandomFileName().Replace(".", string.Empty)
+        };
+        // Replace first character with a random letter if it's a digit to avoid random insertions of '_'
```
Should we just write our own ProjectName name generator from a smaller set of chars, rather than relying on Guid.NewGuid, so we don't have to do this funky biz here?

HaoK replied:
or alternatively just jam a random letter in front of the guid on creation time

DamianEdwards replied:
We could, although this is using `Path.GetRandomFileName()` rather than `Guid.NewGuid` here.

DamianEdwards replied:
I'll change this to append the letter to the front

HaoK commented on src/ProjectTemplates/test/BlazorServerTemplateTest.cs:
```diff
@@ -24,18 +25,18 @@ public BlazorServerTemplateTest(ProjectFactoryFixture projectFactory)
+    [InlineData("Individual", null)]
+    [InlineData("Individual", new string[] { ArgConstants.UseLocalDb })]
+    [InlineData("Individual", new string[] { ArgConstants.UseProgramMain })]
+    [InlineData("Individual", new string[] { ArgConstants.UseProgramMain, ArgConstants.UseLocalDb })]
```
nit, could drop the "Individual" and just hardcode that in CreateBuildPublishAsync

DamianEdwards replied:
The auth is passed through from the other theory though too, `BlazorServerTemplate_IdentityWeb_BuildAndPublish`

HaoK replied:
Sorry I just meant drop it from the inline data, and call `CreateBuildPublishAsync("Individual", args: args);` or am I misreading this

DamianEdwards replied:
Yeah doesn't work: 
<img width="842" alt="image" src="https://user-images.githubusercontent.com/249088/167743478-89d402b6-1cb6-4818-b1e2-67fdcca53189.png">

HaoK replied:
Weird oh well

DamianEdwards committed:
Fix xunit warning

MarcoRossignoli commented:
cc: @nohwnd

HaoK committed:
Bump test timeout to 60m to see if that helps

DamianEdwards committed:
Print test timeout message & prepare for clean-up

DamianEdwards committed:
Ignore process exit code when printing CTS timeout

nohwnd commented:
I had a look on your logs, and there is no indication of what is happening, other than that a template test always runs when crash happens. Testhost seems to abruptly die. 

So there are few things you can do: 

- Try enabling --blame-crash to see if you can capture a crash dump.
- If the crash happens on Windows, I would try to get few of the last Application Error logs from Event Log they have useful info.
```
get-eventlog -LogName application -entrytype error -Source "application error" | Select -First 10 | Format-List
```
- Are you by any chance connecting the processes into a Job (group of processes), so when one process dies the others follow?

DamianEdwards committed:
Add timestamps to helix runner console logs & bump helix timeout for project template tests

DamianEdwards committed:
Skip LocalDb template tests on non-Windows platforms

HaoK commented on src/ProjectTemplates/test/BlazorServerTemplateTest.cs:
```diff
@@ -24,18 +25,23 @@ public BlazorServerTemplateTest(ProjectFactoryFixture projectFactory)
+    [Theory]
+    [InlineData("Individual", new string[] { ArgConstants.UseLocalDb })]
+    [InlineData("Individual", new string[] { ArgConstants.UseProgramMain, ArgConstants.UseLocalDb })]
+    [SkipOnHelix("https://github.com/dotnet/aspnetcore/issues/30825", Queues = "All.OSX")]
```
nit: can remove this Skip on All.OSX now that these variants are windows only

DamianEdwards committed:
Actually skip tests

HaoK commented:
@nohwnd so what we are trying in this PR is: 

`var commonTestArgs = $"test {Options.Target} --diag:{diagLog} --logger:xunit --logger:\"console;verbosity=normal\" --blame \"CollectHangDump;CollectDump;TestTimeout=15m\"";`

It looks like some new blame variants have been added like the one you suggested, can we drop our blame with CollectHangDump/Dump/TestTimeout and replace with blame-crash and blame-hang-timeout:

`var commonTestArgs = $"test {Options.Target} --diag:{diagLog} --logger:xunit --logger:\"console;verbosity=normal\" --blame-crash --blame-hang-timeout \"15m\"";`

DamianEdwards committed:
Update test cmd args in TestRunner.cs

DamianEdwards commented:
@HaoK @nohwnd I've updated it here https://github.com/dotnet/aspnetcore/pull/41563/commits/3af73524200bb62bca7101fce0cd1e9e6d55c025

DamianEdwards committed:
Split up long-running theory to avoid test timeouts

nohwnd commented:
@HaoK the newer --blame-crash syntax translates to what you already had. I did not see you are already using CollectDump in the code, I probably overlooked it or saw the code before it was added. 

---

In the latest [build with abort](https://dev.azure.com/dnceng/public/_build/results?buildId=1766073&view=logs&j=8c9dd6ae-5017-5a70-997c-c2386fed4e85&t=e6acbd31-8469-5d42-602d-cb75e71f3b60), the vstest logs look like processes are killed externally, and this log indicates that this is the case: 

2022-05-12T00:27:28.8279660Z Non-quarantined tests exceeded configured timeout: 40m.

So I guess you have one more timeout to change? :)

DamianEdwards committed:
Remove template test locks & skip template tests on Debian11Arm

- Deleted unused code
- Updated README with details about test timeouts & conditional tests

DamianEdwards committed:
Revert helix matrix config

BrennanConroy commented on eng/tools/HelixTestRunner/TestRunner.cs:
```diff
@@ -22,33 +22,35 @@ public bool SetupEnvironment()
+            ProcessUtil.PrintMessage($"Set VSTEST_DUMP_PATH: {dumpPath}");
             EnvironmentVariables.Add("VSTEST_DUMP_PATH", dumpPath);
+            EnvironmentVariables.Add("DOTNET_CLI_VSTEST_TRACE", "1");
+            EnvironmentVariables.Add("VSTEST_ DISABLE_FASTER_JSON_SERIALIZATION=", "1");
```
Is this option doing anything? There is a space in the middle of it and it includes a `=` at the end which isn't normal

DamianEdwards replied:
lol, hmm, I guess not, I'll remove

HaoK reviewed (approved):
Thanks for grinding through this even after falling into all the pits of doom, things look way better!

DamianEdwards committed:
PR feedback

BrennanConroy commented on src/ProjectTemplates/README.md:
```diff
@@ -62,6 +62,101 @@ Then, use one of:
 previous step, it is NOT advised that you install templates created on your local machine using just
 `dotnet new -i [nupkgPath]`.
 
+#### Conditional tests & skipping test platforms
```
How much of this is specific to template tests? Should it be documented in another location instead?

DamianEdwards replied:
Yeah @HaoK said the same thing in a comment on a commit before I rebased (so it got lost from here) but I'll do that shuffle once this is in main.

dougbu reviewed (changes requested):
One thing to clean up and a few questions&hellip;

DamianEdwards commented on src/ProjectTemplates/test/ArgConstants.cs:
```diff
@@ -23,4 +23,5 @@ internal static class ArgConstants
     public const string AppIdClientId = "--api-client-id";
     public const string TenantId = "--tenant-id";
     public const string AadB2cInstance = "--aad-b2c-instance";
+    public const string UseLocalDb = "-uld";
```
This is what was already being used where it was a string literal so I just preserved it.

DamianEdwards commented on src/ProjectTemplates/Shared/ProjectFactoryFixture.cs:
```diff
@@ -24,9 +24,25 @@ public ProjectFactoryFixture(IMessageSink diagnosticsMessageSink)
+        return project;
+    }
+
     public async Task<Project> GetOrCreateProject(string projectKey, ITestOutputHelper output)
```
It's in the shared directory so could be used outside of these tests but I didn't check if it's used anywhere else

DamianEdwards replied:
Seems it can be deleted.

DamianEdwards commented on src/ProjectTemplates/test/BlazorServerTemplateTest.cs:
```diff
@@ -24,31 +25,36 @@ public BlazorServerTemplateTest(ProjectFactoryFixture projectFactory)
+    [ConditionalTheory]
+    [InlineData("Individual", new string[] { ArgConstants.UseLocalDb })]
+    [InlineData("Individual", new string[] { ArgConstants.UseProgramMain, ArgConstants.UseLocalDb })]
+    [OSSkipCondition(OperatingSystems.Linux | OperatingSystems.MacOSX, SkipReason = "No LocalDb on non-Windows")]
```
The pages don't fail unless you try to actually register a user (i.e. hit the database).

DamianEdwards commented on src/ProjectTemplates/test/ProjectTemplates.Tests.csproj:
```diff
@@ -46,29 +49,14 @@
-      Private="false"
-      ReferenceOutputAssembly="false"
-      SkipGetTargetFrameworkProperties="true" />
+    <ProjectReference Include="$(RepoRoot)src\Framework\App.Runtime\src\Microsoft.AspNetCore.App.Runtime.csproj" Private="false" ReferenceOutputAssembly="false" SkipGetTargetFrameworkProperties="true" />
```
VS did that automatically so you're fighting against it if we keep changing this back.

DamianEdwards replied:
Reverted this formatting

DamianEdwards commented on src/ProjectTemplates/test/BlazorServerTemplateTest.cs:
```diff
@@ -3,6 +3,7 @@
 
 using System;
 using System.Collections.Generic;
+using System.Globalization;
```
I think there was a `CultureInfo` in here at one point but it seems it's gone now.

DamianEdwards committed:
PR feedback
